File: app/static/setup-proposal.js
/* Exact, read-only setup review. No mutation, upload or execution interface. */
(function(root,factory){const api=factory();if(typeof module==='object'&&module.exports)module.exports=api;if(root)root.StudioSetupProposal=api;})(typeof window!=='undefined'?window:null,function(){
  'use strict';
  const FORMAT='studio.setup-proposal/v1',API='/api/workflow-studio/setup-proposal',HASH=/^[0-9a-f]{64}$/;
  const clone=x=>JSON.parse(JSON.stringify(x));
  function need(ok,message){if(!ok)throw Error(message);}
  function stable(x){if(Array.isArray(x))return '['+x.map(stable).join(',')+']';if(x&&typeof x==='object')return '{'+Object.keys(x).sort().map(k=>JSON.stringify(k)+':'+stable(x[k])).join(',')+'}';return JSON.stringify(x);}
  const same=(a,b)=>stable(a)===stable(b);
  function expectedDiff(before,intent){const r=before.recipe;return [
    ['Recipe',{preset:r.preset,template:before.templateHash},{preset:intent.preset_id,template:intent.template_sha256}],
    ['Settings and wording',r.controls,intent.controls],['Batch',r.batch,intent.batch],['References',r.references,intent.sources],
    ['Lineage',{parents:r.parent_assets,by_input:r.parent_by_input},intent.lineage],
    ['Continuation context',r.continuation??null,null],['Pending local inputs',before.pendingInputs,[]]
  ].map(([section,before,proposed])=>({section,before,proposed,changed:!same(before,proposed)}));}
  async function sha256(text){need(globalThis.crypto?.subtle,'Exact review requires a secure browser context. Open Studio on its configured loopback address.');return Array.from(new Uint8Array(await crypto.subtle.digest('SHA-256',new TextEncoder().encode(text))),x=>x.toString(16).padStart(2,'0')).join('');}
  async function validate(r,q,hashText=sha256){
    const message='Setup proposal response does not match the captured request.';
    need(r&&r.format===FORMAT&&r.can_apply===false&&r.generation_submitted===false&&r.execution_authorized===false,message);
    need(same(r.request,q)&&same(r.before,q.draft)&&r.precondition?.scope==='caller-declared browser draft; not a server revision'&&HASH.test(r.precondition?.draft_sha256||''),message);
    need(typeof r.proposal_json==='string'&&new TextEncoder().encode(r.proposal_json).length<=1048576&&HASH.test(r.proposal_sha256||''),message);
    const {proposal_json,proposal_sha256,...body}=r;
    need(same(JSON.parse(proposal_json),body)&&await hashText(proposal_json)===proposal_sha256,message);
    need(r.intent?.preset_id===q.preset_id&&r.intent?.template_sha256===q.expected_template_sha256&&r.intent.batch===1,message);
    need(r.intent.controls&&typeof r.intent.controls==='object'&&!Array.isArray(r.intent.controls)&&!['reference','last_reference'].some(k=>Object.hasOwn(r.intent.controls,k)),message);
    for(const key of ['positive','negative'])need((r.intent.controls[key]??'')===q[key],message);
    need(Array.isArray(r.intent.sources)&&r.intent.sources.length===q.sources.length&&r.intent.sources.every((x,i)=>x?.asset_id===q.sources[i].asset_id&&x.sha256===q.sources[i].sha256&&x.role===q.sources[i].role&&x.slot===i+1&&x.staged===false&&Array.isArray(x.binding)&&x.binding.length===2&&x.binding[1]==='image'&&['contribution','avoid'].every(k=>x[k]===q.guidance[i][k].trim())),message);
    need(same(r.diff,expectedDiff(q.draft,r.intent))&&Array.isArray(r.side_effects)&&Array.isArray(r.limits)&&Array.isArray(r.observation?.candidate?.checks),message);
    return r;
  }
  class Session{
    constructor(transport,matches,emit,options={}){this.transport=transport;this.matches=matches;this.emit=emit;this.hashText=options.hashText||sha256;this.set=options.set||(fn=>setTimeout(fn,15000));this.clear=options.clear||(id=>clearTimeout(id));this.epoch=0;this.busy=false;this.current=null;}
    invalidate(message='Proposal inputs changed. Build a new preview.'){this.epoch++;const a=this.current;this.current=null;this.busy=false;if(a){this.clear(a.timer);a.controller.abort();}this.emit({busy:false,message});}
    async load(value){
      if(this.busy)return;const q=clone(value),epoch=++this.epoch,active={controller:new AbortController(),timer:null};this.current=active;this.busy=true;this.emit({busy:true,message:'Building a review-only proposal…'});
      active.timer=this.set(()=>{if(this.current===active)this.invalidate('The preview timed out. No proposal was accepted. Retry explicitly.');});
      const current=()=>epoch===this.epoch&&this.current===active;
      try{const response=await this.transport(q,active.controller.signal);if(!current())return;need(this.matches(),'The current draft changed. Close this preview and check recipes again.');const report=await validate(response,q,this.hashText);if(!current())return;need(this.matches(),'The draft changed while the proposal was checked. Reopen the chooser.');this.busy=false;this.emit({busy:false,report});}
      catch(error){if(current()){this.busy=false;this.emit({busy:false,message:String(error?.message||error)});}}
      finally{this.clear(active.timer);if(this.current===active){this.current=null;this.busy=false;}}
    }
  }
  function mount(w,options={}){
    const d=w.document,$=s=>d.querySelector(s);if(!$('#createView')||$('#setupProposalDialog'))return;
    function el(tag,text,cls){const n=d.createElement(tag);if(text!==undefined)n.textContent=text;if(cls)n.className=cls;return n;}
    const dialog=el('dialog',undefined,'setup-proposal');dialog.id='setupProposalDialog';dialog.setAttribute('aria-labelledby','setupProposalTitle');
    const title=el('h2','Review a proposed setup');title.id='setupProposalTitle';
    const close=el('button','Close preview');close.type='button';close.id='closeSetupProposal';
    const heading=el('div',undefined,'proposal-heading');heading.append(title,close);
    const intro=el('p','Nothing is applied here. Review the target defaults, wording and every source before a future setup handoff. Your current Create setup stays untouched.');
    const summary=el('p',undefined,'muted');summary.id='setupProposalSummary';
    const inputs=el('details'),inputsTitle=el('summary','Wording and source contributions');inputs.id='setupProposalInputs';inputs.open=true;inputs.append(inputsTitle);
    const form=el('form'),positive=el('textarea'),negative=el('textarea'),sources=el('div');
    positive.id='proposalPositive';negative.id='proposalNegative';positive.maxLength=negative.maxLength=8000;positive.rows=3;negative.rows=2;
    function field(parent,labelText,input){const label=el('label',labelText);label.htmlFor=input.id;parent.append(label,input);}
    field(form,'Desired result — starts from your current wording, not a recipe example',positive);field(form,'Negative wording — clear this for a route without a negative-prompt input',negative);
    form.append(sources);const build=el('button','Build proposal');build.id='buildSetupProposal';build.type='submit';form.append(build);
    const status=el('p','Review wording and source contributions, then build the proposal.');status.id='setupProposalStatus';status.setAttribute('role','status');status.setAttribute('aria-live','polite');
    const output=el('div');output.id='setupProposalResult';const download=el('button','Export proposal JSON');download.id='exportSetupProposal';download.type='button';download.disabled=true;
    inputs.append(form);dialog.append(heading,intro,summary,inputs,status,output,download,el('p','Export contains your captured prompt, reference metadata and proposed settings. No image bytes are included.','muted'));d.body.append(dialog);
    let context=null,before=null,stamp=null,opener=null,report=null,stale=false;
    const fields=[];
    const matches=()=>!!context&&!stale&&stamp===w.StudioSetupDraft?.stamp();
    const session=new Session(async(q,signal)=>{
      const response=await w.fetch(API,{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify(q),signal});const text=await response.text();need(new TextEncoder().encode(text).length<=2*1048576,'Proposal exceeds the browser response limit.');
      let value;try{value=JSON.parse(text);}catch(_){throw Error('The preview response was unreadable. Retry explicitly.');}need(response.ok,typeof value?.error==='string'?value.error:'Setup preview unavailable.');return value;
    },matches,event=>{report=null;output.replaceChildren();download.disabled=true;build.disabled=event.busy||stale;dialog.setAttribute('aria-busy',String(event.busy));if(event.report){report=event.report;render(report);download.disabled=false;}else status.textContent=event.message;},options);
    function invalidate(message,contextChanged=false){if(!dialog.open)return;if(contextChanged)stale=true;session.invalidate(message);}
    function render(r){
      status.textContent='Proposal ready for review only. No settings, files or jobs were changed.';
      inputs.open=false;inputsTitle.textContent='Edit wording and source contributions';const resultTitle=el('h3','What would change');resultTitle.tabIndex=-1;output.append(resultTitle);
      for(const row of r.diff){const detail=el('details',undefined,'proposal-change');detail.open=row.changed&&['Recipe','Settings and wording','References','Batch','Pending local inputs'].includes(row.section);detail.append(el('summary',row.section+' · '+(row.changed?'would change':'unchanged')));
        if(row.section==='Settings and wording'){
          const table=el('table'),head=el('tr');for(const label of ['Control','Current','Proposed'])head.append(el('th',label));const thead=el('thead');thead.append(head);table.append(thead);const body=el('tbody');
          for(const key of Array.from(new Set([...Object.keys(row.before),...Object.keys(row.proposed)])).sort()){const tr=el('tr');tr.append(el('th',key),el('td',Object.hasOwn(row.before,key)?String(row.before[key]):'Not present'),el('td',Object.hasOwn(row.proposed,key)?String(row.proposed[key]):'Would be removed'));body.append(tr);}table.append(body);detail.append(table);
        }else if(row.section==='References'){
          detail.append(el('p',row.before.length+' current role reference(s) would be replaced. Any filename-based inputs are listed in Settings and wording.'));
          for(const item of row.proposed){const card=el('section',undefined,'proposal-source');card.append(el('h4','Picture '+item.slot+' · '+item.title+' · '+item.role),el('p','Use: '+(item.contribution||'the assigned visual role')+' · Do not transfer: '+(item.avoid||'unrequested details')));
            const t=item.transform;card.append(el('p',item.width+' × '+item.height+' source pixels · '+t.policy+(t.resized_size?' → '+t.resized_size.join(' × '):'')));
            const exact=el('details');exact.append(el('summary','Exact source, slot and transform'),el('pre',JSON.stringify(item,null,2)));card.append(exact);detail.append(card);}
        }else{const columns=el('div',undefined,'proposal-columns');for(const [label,value]of [['Current',row.before],['Proposed',row.proposed]]){const box=el('section');box.append(el('h4',label),el('pre',JSON.stringify(value,null,2)));columns.append(box);}detail.append(columns);}output.append(detail);
      }
      if(r.intent.compiled_positive!==null){const detail=el('details');detail.append(el('summary','Exact primary prompt after role guidance'),el('pre',r.intent.compiled_positive),el('p','This uses the existing reference compiler. Any other positive bindings receive the unprefixed desired-result text.'));output.append(detail);}
      const holds=el('details');holds.append(el('summary','Prerequisites and remaining checks'));for(const item of r.observation.candidate.checks)holds.append(el('p',item.message));output.append(holds);
      const boundaries=el('details');boundaries.append(el('summary','Side effects and limits'));for(const note of [...r.side_effects,...r.limits])boundaries.append(el('p',note,'muted'));output.append(boundaries);
      const identity=el('details');identity.append(el('summary','Proposal identity and draft boundary'),el('code',r.proposal_sha256),el('p',r.precondition.scope));output.append(identity);const apply=w.StudioSetupApply?.offer?.(r,()=>dialog.open&&report===r&&matches());if(apply)output.append(apply);dialog.scrollTop=0;resultTitle.focus({preventScroll:true});
    }
    function open(event){
      try{
        const detail=event.detail;need(detail&&Array.isArray(detail.sources)&&detail.sources.length>=1&&detail.sources.length<=3,'Choose exact source images before previewing a setup.');
        need(w.StudioSetupDraft,'The Create draft is not available yet.');before=w.StudioSetupDraft.capture();need(before,'Select a current Create recipe before reviewing a replacement.');
        need(!w.StudioSetupDraft.busy(),'Wait for the current reference attachment to settle before previewing.');
        context=clone(detail);stamp=w.StudioSetupDraft.stamp();stale=false;opener=d.activeElement;report=null;
        positive.value=before.recipe.controls.positive||'';negative.value=before.recipe.controls.negative||'';fields.length=0;sources.replaceChildren();
        summary.textContent='Target: '+context.preset_id+' · '+context.sources.length+' source image(s). Sampling and adapter settings start from the target graph defaults; the comparison lists every removed current control.';
        context.sources.forEach((source,index)=>{const box=el('fieldset');box.append(el('legend','Picture '+(index+1)+' · '+source.role));const contribution=el('textarea'),avoid=el('textarea');contribution.id='proposalContribution'+(index+1);avoid.id='proposalAvoid'+(index+1);contribution.rows=avoid.rows=2;contribution.maxLength=avoid.maxLength=1500;
          field(box,'Use from this image',contribution);field(box,'Do not transfer',avoid);box.append(el('small','Workspace identity: '+source.asset_id));fields.push({contribution,avoid});sources.append(box);});
        inputs.open=true;inputsTitle.textContent='Wording and source contributions';session.invalidate('Review wording and contributions, then build the proposal.');if(!dialog.open)dialog.showModal();positive.focus();
      }catch(error){const message=$('#shortlistStatus');if(message)message.textContent=error.message;}
    }
    form.onsubmit=e=>{e.preventDefault();try{need(matches(),'The draft or recipe advice changed. Close this preview and check again.');need(!w.StudioSetupDraft.busy(),'A reference attachment is in progress. Reopen after it settles.');session.load({...context,draft:before,positive:positive.value,negative:negative.value,guidance:fields.map(x=>({contribution:x.contribution.value,avoid:x.avoid.value}))});}catch(error){invalidate(error.message,true);}};
    form.addEventListener('input',()=>invalidate('Wording or contributions changed. Build a new proposal.'));
    download.onclick=()=>{try{need(report&&matches(),'The draft changed; this proposal cannot be exported as current.');const url=URL.createObjectURL(new Blob([JSON.stringify(report,null,2)],{type:'application/json'})),link=el('a');link.href=url;link.download='setup-proposal-'+report.proposal_sha256.slice(0,12)+'.json';link.click();setTimeout(()=>URL.revokeObjectURL(url),1000);}catch(error){invalidate(error.message,true);}};
    close.onclick=()=>dialog.close();dialog.addEventListener('close',()=>{session.invalidate('Preview closed. Any explicit setup operation is reported in Shared setup & recovery.');if(opener?.isConnected)opener.focus();else $('#checkStartingRecipes')?.focus();});
    d.addEventListener('studio:setup-proposal',open);
    d.addEventListener('studio:shortlist-invalidated',()=>invalidate('Recipe advice changed. Close this preview and check again.',true));
    d.addEventListener('studio:recipe',()=>invalidate('The Create setup changed. Close this preview and check again.',true));
    for(const name of ['input','change'])$('#createView').addEventListener(name,()=>invalidate('The Create inputs changed. Close this preview and check again.',true));
    w.addEventListener('pagehide',()=>{session.invalidate('Preview closed.');if(dialog.open)dialog.close();});
    w.addEventListener('hashchange',()=>{if(w.location.hash!=='#create'&&dialog.open)dialog.close();});
  }
  return{validate,Session,same,mount};
});

File: studio_workflow/setup_proposal.py
"""Review-only setup intents over the existing shortlist and reference compiler.

A captured browser draft is a caller declaration, not a shared revision. These
reports deliberately cannot be submitted as saved recipes or execution tickets.
"""
from __future__ import annotations

import argparse
import copy
import hashlib
import json
import math
import re
import sys
from http.client import HTTPException
from pathlib import Path
from urllib.error import HTTPError

from .core import canonical, decode, digest, need
from .shortlist import query as shortlist_query, request as shortlist_request, _runtime

PREFIX='/api/workflow-studio/setup-proposal'
FORMAT='studio.setup-proposal/v1'
MAX_BYTES=131072
DRAFT_SCOPE='caller-declared browser draft; not a server revision'
HASH=re.compile(r'[0-9a-f]{64}\Z')
ID=re.compile(r'[A-Za-z0-9_.-]{1,96}\Z')


def _text(value, maximum):
    return type(value) is str and len(value)<=maximum


def _exact_json(value):
    """Copy bounded JSON; callers must use text for integers unsafe in a browser."""
    try:
        raw=canonical(value)
        need(len(raw)<=MAX_BYTES, 'Setup proposal input exceeds 128 KiB')
        result=decode(raw)
    except (TypeError, OverflowError, RecursionError) as exc:
        raise ValueError('Supply bounded JSON setup data') from exc
    def check(item):
        if type(item) in (int,float): need(abs(item)<=2**53-1, 'Use text for wide integers to preserve browser precision')
        elif type(item) is list:
            for x in item: check(x)
        elif type(item) is dict:
            for x in item.values(): check(x)
    check(result)
    return result


def query(value):
    q=_exact_json(value)
    need(type(q) is dict and set(q)=={'goal','preset_id','expected_template_sha256','sources','draft','positive','negative','guidance'},
         'Supply goal, preset_id, expected_template_sha256, ordered sources, draft, positive, negative and guidance')
    need(type(q['preset_id']) is str and ID.fullmatch(q['preset_id']), 'Invalid target preset identity')
    need(type(q['expected_template_sha256']) is str and HASH.fullmatch(q['expected_template_sha256']), 'Invalid target graph identity')
    shortlist_query({'goal':q['goal'],'sources':q['sources']})
    need(_text(q['positive'],8000) and _text(q['negative'],8000), 'positive and negative wording must be text up to 8000 characters')
    need(type(q['guidance']) is list and len(q['guidance'])==len(q['sources']), 'Provide guidance for every Picture in order')
    for row in q['guidance']:
        need(type(row) is dict and set(row)=={'contribution','avoid'} and all(_text(x,1500) for x in row.values()),
             'Reference guidance must contain contribution and avoid text up to 1500 characters')
    q['draft']=validate_draft(q['draft'])
    return q


def validate_draft(value):
    """Shared normalized Create data; no persistence or staging authority."""
    draft=_exact_json(value)
    need(type(draft) is dict and set(draft)=={'version','updatedAt','recipe','pendingInputs','templateHash'}
         and type(draft['version']) is int and draft['version']==1 and type(draft['updatedAt']) is int and draft['updatedAt']==0,
         'Supply a complete normalized browser draft with updatedAt=0')
    need(draft['templateHash'] is None or type(draft['templateHash']) is str and HASH.fullmatch(draft['templateHash']), 'Invalid draft template identity')
    need(type(draft['pendingInputs']) is list and len(draft['pendingInputs'])<=2
         and all(x in ('reference','lastReference') for x in draft['pendingInputs']) and len(set(draft['pendingInputs']))==len(draft['pendingInputs']),
         'Invalid draft pending inputs')
    recipe=draft['recipe']; required={'preset','controls','batch','references','parent_assets','parent_by_input'}
    need(type(recipe) is dict and required<=set(recipe) and set(recipe)<=required|{'continuation'}, 'Supply the complete current recipe declaration')
    need(type(recipe['preset']) is str and ID.fullmatch(recipe['preset']), 'Invalid draft preset')
    need(type(recipe['batch']) is int and 1<=recipe['batch']<=4, 'Invalid draft batch')
    need(type(recipe['controls']) is dict and len(recipe['controls'])<=64, 'Invalid draft controls')
    for key,item in recipe['controls'].items():
        need(re.fullmatch(r'[a-z][a-z0-9_]{0,63}',key) and (_text(item,20000) or type(item) in (int,float)), 'Invalid draft control')
    need(type(recipe['references']) is list and len(recipe['references'])<=8 and all(type(x) is dict for x in recipe['references']), 'Invalid retained reference declarations')
    need(type(recipe['parent_assets']) is list and len(recipe['parent_assets'])<=32 and all(_text(x,160) for x in recipe['parent_assets']), 'Invalid draft lineage')
    need(type(recipe['parent_by_input']) is dict and set(recipe['parent_by_input'])<={'reference','lastReference'}
         and all(type(x) is str and x in recipe['parent_assets'] for x in recipe['parent_by_input'].values()), 'Invalid draft input lineage')
    need('continuation' not in recipe or recipe['continuation'] is None or type(recipe['continuation']) is dict, 'Invalid draft continuation')
    return draft


def _capture(studio, items):
    from app.references import image_record
    from PIL import Image
    result=[]
    for index,item in enumerate(items):
        label=f'Picture {index+1}: '
        try:asset=studio.assets.get(item['asset_id'])
        except (OSError,ValueError,KeyError) as exc:raise ValueError(label+'Workspace image unavailable; reselect it.') from exc
        need(asset and not asset.get('trashed_at') and asset.get('media_type')=='image', label+'Restore an available image from the Workspace before previewing.')
        try:
            path=studio.assets.file(item['asset_id'])
            record=image_record(path.parent,path.name,strict_pixels=True)
        except (OSError,ValueError,KeyError,Image.DecompressionBombError,Image.DecompressionBombWarning) as exc:
            raise ValueError(label+'Image could not be read within the byte/pixel safety limits; choose a valid smaller image.') from exc
        need(record['sha256']==item['sha256']==asset['sha256'], label+'Source bytes changed; select the intended image again.')
        result.append({**item,'slot':index+1,'title':str(asset.get('title') or item['asset_id'])[:160],
                       **{k:record[k] for k in ('bytes','width','height')},'staged':False})
    return result


def _controls(preset, graph, positive, negative):
    from app.server import CONTROL_KEYS
    controls={};bindings={}
    for key in CONTROL_KEYS:
        if key in ('reference','last_reference'): continue
        targets=([preset[key]] if preset.get(key) else [])+list((preset.get('bindings_extra') or {}).get(key,[]))
        if not targets:
            need(key not in ('positive','negative') or not (positive if key=='positive' else negative), f'This route has no {key} prompt binding; wording would be dropped')
            continue
        values=[]
        for target in targets:
            need(type(target) is list and len(target)==2, 'Unsupported control binding: '+key)
            node,field=map(str,target)
            need(node in graph and type(graph[node].get('inputs')) is dict and field in graph[node]['inputs'], 'Missing control binding: '+key)
            value=graph[node]['inputs'][field]
            need(type(value) in (str,int,float) and (type(value) is not float or math.isfinite(value)), 'Unsupported default control: '+key)
            values.append(value)
        if key in ('positive','negative'): value=positive if key=='positive' else negative
        else:
            need(all(canonical(x)==canonical(values[0]) for x in values), 'Default fan-out differs for '+key+'; inspect it in the existing editor')
            value=str(values[0])
        need(len(value)<=8000, 'Default control exceeds review limit: '+key)
        controls[key]=value;bindings[key]=copy.deepcopy(targets)
    return controls,bindings


def _diff(before, intent):
    recipe=before['recipe']
    sections=[('Recipe',{'preset':recipe['preset'],'template':before['templateHash']},{'preset':intent['preset_id'],'template':intent['template_sha256']}),
              ('Settings and wording',recipe['controls'],intent['controls']),('Batch',recipe['batch'],intent['batch']),
              ('References',recipe['references'],intent['sources']),
              ('Lineage',{'parents':recipe['parent_assets'],'by_input':recipe['parent_by_input']},intent['lineage']),
              ('Continuation context',recipe.get('continuation'),None),('Pending local inputs',before['pendingInputs'],[])]
    return [{'section':title,'before':old,'proposed':new,'changed':canonical(old)!=canonical(new)} for title,old,new in sections]


def request(value, studio):
    q=query(value)
    runtime=_runtime(studio);catalog_bytes=studio.catalog_path.read_bytes()
    need(len(catalog_bytes)<=8*1024*1024, 'Preset catalog exceeds review limit')
    entries=studio.catalog()['presets'];matches=[p for p in entries if p.get('id')==q['preset_id']]
    need(len(matches)==1, 'Target recipe is unavailable or ambiguous')
    preset=copy.deepcopy(matches[0]);graph,path=studio.graph_for(preset);graph=copy.deepcopy(graph)
    raw=path.read_bytes();need(len(raw)<=1048576 and hashlib.sha256(raw).hexdigest()==q['expected_template_sha256'], 'Target graph identity changed; check recipes again')
    need(canonical(graph)==canonical(decode(raw)), 'Parsed graph differs from the verified target bytes; check again')
    need(preset.get('continuation_capability',{}).get('template_sha256')==q['expected_template_sha256'], 'Target graph identity differs from catalog')
    captured=_capture(studio,q['sources'])
    report=shortlist_request({'goal':q['goal'],'sources':q['sources'],'limit':1},studio,preset_id=q['preset_id'])
    need(len(report['candidates'])==1 and report['total']==1, 'The target recipe does not match this goal')
    candidate=report['candidates'][0];assignments=candidate.get('source_assignments',[])
    from app.references import reference_transform, guidance_text, board_spec, board_transform, ROLES
    named=preset.get('reference_slots') or [];board=board_spec(preset) if preset.get('reference_board') else None
    if board:
        need(board['minimum']<=len(captured)<=board['slot_count'],
             f'This style board needs {board["minimum"]}-{board["slot_count"]} ordered Picture sources')
        need(len(assignments)==len(captured), 'Every supplied board Picture needs a distinct target slot')
    else:
        need(len(assignments)==len(captured) and candidate['reference_count']==len(captured), 'Every Picture needs a distinct target slot')
    need(not preset.get('requires_rgba_mask') and not preset.get('continuation_capability',{}).get('requires_mask'), 'Mask routes need a reviewed mask; setup proposal does not validate masks')
    # Motion modes and native transforms require a separate complete proposed-state contract.
    need(not preset.get('i2v_modes'), 'Motion-mode proposals are not supported yet; use the existing animation controls')
    controls,bindings=_controls(preset,graph,q['positive'],q['negative'])
    source_intents=[]
    for index,(source,assignment,guidance) in enumerate(zip(captured,assignments,q['guidance'])):
        need(assignment.get('binding') is not None, f'Picture {index+1} has no distinct target binding')
        if board:
            need(not guidance['contribution'] and not guidance['avoid'],
                 f'Picture {index+1}: style-board images are visual-only; per-Picture prompt guidance is not compiled')
            need(source['role']==board['roles'][index], f'Picture {index+1}: choose the {board["roles"][index]} board role before previewing')
            role_mode='style-board';transform=board_transform(preset);binding=copy.deepcopy(named[index]['binding'])
        elif named:
            need(source['role'] in ROLES and assignment.get('role_mode')=='prompt-guidance', f'Picture {index+1}: choose a named reference role before previewing wording')
            role_mode=assignment['role_mode'];binding=assignment['binding']
            transform=reference_transform(preset,graph,binding[0],source)
        else:
            need(assignment.get('role_mode')=='whole-image', f'Picture {index+1} has no supported whole-image binding')
            need(not guidance['contribution'] and not guidance['avoid'], 'This whole-image route does not compile per-Picture guidance; review the main wording instead')
            role_mode=assignment['role_mode'];binding=assignment['binding']
            transform={'policy':'native graph preprocessing; not projected by this adapter'}
        source_intents.append({**source,**{k:v.strip() for k,v in guidance.items()},'binding':binding,
                               'role_mode':role_mode,'transform':transform})
    compiled=controls.get('positive') if board or not named else guidance_text(source_intents,q['positive'])
    by_input={} if board else {'reference':captured[0]['asset_id']}
    intent={'preset_id':q['preset_id'],'template_sha256':q['expected_template_sha256'],'backend_id':candidate['backend_id'],
            'controls':controls,'control_bindings':bindings,'batch':1,'sources':source_intents,'compiled_positive':compiled,
            'compiled_positive_binding':preset.get('positive'),
            'lineage':{'parents':list(dict.fromkeys(x['asset_id'] for x in captured)),'by_input':by_input,
                       'source_slots':[{'slot':x['slot'],'asset_id':x['asset_id'],'role':x['role']} for x in captured]}}
    if board:intent['reference_board']={k:board[k] for k in ('minimum','slot_count','roles')}
    if not named and len(captured)==2 and assignments[1]['binding']==preset.get('last_reference'):
        intent['lineage']['by_input']['lastReference']=captured[1]['asset_id']
    # These are observations; no leases, mutations or preflight/worker calls occur.
    need(captured==_capture(studio,q['sources']), 'Source context changed during preview; check again')
    need(runtime==_runtime(studio) and catalog_bytes==studio.catalog_path.read_bytes() and raw==path.read_bytes(), 'Studio context changed during preview; check again')
    result={'format':FORMAT,'request':q,'before':q['draft'],'intent':intent,'diff':_diff(q['draft'],intent),
            'precondition':{'draft_sha256':digest(q['draft']),'scope':DRAFT_SCOPE},
            'observation':{'snapshot_sha256':report['snapshot_sha256'],'backend_id':report['backend_id'],
                           'schema_sha256':report['schema_sha256'],'candidate':candidate},
            'side_effects':['Preview performs no attachment, source upload, setting change, save or generation.',
                            'A future apply would replace the listed setup, reset batch to one, clear continuation context and require explicit source staging.',
                            'Pending local inputs and old reference/lineage declarations must be retained for a separately implemented undo.'],
            'limits':['Draft values were supplied by the caller, not read from an authoritative shared draft.',
                      'Reference geometry uses the existing compiler projection; native interpolation, masks, licensing and artistic quality are not certified.',
                      'Prerequisite observations concern defaults; preparation must recheck exact state. This report is not a recipe or run ticket.'],
            'can_apply':False,'execution_authorized':False,'generation_submitted':False}
    need(len(canonical(result))<=1048576, 'Setup proposal response exceeds review limit')
    exact=canonical(result).decode('utf-8')
    return {**result,'proposal_sha256':hashlib.sha256(exact.encode('utf-8')).hexdigest(),'proposal_json':exact}


def validate_reply(result, value):
    q=query(value);message='Setup proposal response does not match the captured request'
    need(type(result) is dict and result.get('format')==FORMAT and canonical(result.get('request'))==canonical(q) and canonical(result.get('before'))==canonical(q['draft']),message)
    need(all(result.get(k) is False for k in ('can_apply','execution_authorized','generation_submitted')),message)
    need(result.get('precondition')=={'draft_sha256':digest(q['draft']),'scope':DRAFT_SCOPE},message)
    need(type(result.get('proposal_json')) is str and len(result['proposal_json'].encode('utf-8'))<=1048576,message)
    need(result.get('proposal_sha256')==hashlib.sha256(result['proposal_json'].encode('utf-8')).hexdigest()
         and decode(result['proposal_json'])=={k:v for k,v in result.items() if k not in ('proposal_sha256','proposal_json')},message)
    intent=result.get('intent');need(type(intent) is dict and intent.get('preset_id')==q['preset_id'] and intent.get('template_sha256')==q['expected_template_sha256'],message)
    need(type(intent.get('controls')) is dict and type(intent.get('batch')) is int and intent['batch']==1
         and not {'reference','last_reference'}&set(intent['controls']),message)
    for field in ('positive','negative'):
        need(intent['controls'].get(field,'')==q[field],message)
    need(type(intent.get('sources')) is list and len(intent['sources'])==len(q['sources']),message)
    for index,(item,expected) in enumerate(zip(intent['sources'],q['sources'])):
        need(type(item) is dict and all(item.get(k)==v for k,v in expected.items()) and type(item.get('slot')) is int and item['slot']==index+1 and item.get('staged') is False,message)
        need(all(item.get(k)==v.strip() for k,v in q['guidance'][index].items()),message)
        need(item.get('role_mode') in ('whole-image','prompt-guidance','style-board') and type(item.get('binding')) is list,message)
    board=intent.get('reference_board')
    if board is not None:
        need(type(board) is dict and set(board)=={'minimum','slot_count','roles'} and type(board['minimum']) is int
             and type(board['slot_count']) is int and 1<=board['minimum']<=len(intent['sources'])<=board['slot_count']
             and type(board['roles']) is list and len(board['roles'])==board['slot_count'],message)
        need(intent.get('compiled_positive')==intent['controls'].get('positive') and intent.get('lineage',{}).get('by_input')=={}
             and all(x.get('role_mode')=='style-board' and not x.get('contribution') and not x.get('avoid') for x in intent['sources']),message)
    need(result.get('diff')==_diff(q['draft'],intent),message)
    return result


def observe(transport, value):
    from .client import ClientError, read_response
    q=query(value)
    try: result=transport(PREFIX,q)
    except HTTPError as exc:
        try:
            with exc: detail=decode(read_response(exc,65536))
        except (ValueError,OSError,HTTPException):detail={}
        message=detail.get('error') if type(detail) is dict else None
        raise ClientError(exc.code,{'error':message if type(message) is str else 'Setup preview unavailable; retry explicitly.',
                                    'code':'setup_proposal_unavailable','generation_submitted':False}) from exc
    return validate_reply(result,q)


def main(argv=None):
    parser=argparse.ArgumentParser(description='Review a complete setup proposal; never apply, stage or run it.')
    parser.add_argument('request_json',type=Path,help='Strict UTF-8 request JSON, maximum 128 KiB')
    parser.add_argument('--url',default='http://127.0.0.1:8191')
    args=parser.parse_args(argv)
    try:
        with args.request_json.open('rb') as stream:raw=stream.read(MAX_BYTES+1)
        need(len(raw)<=MAX_BYTES,'Setup proposal input exceeds 128 KiB')
        from .sdk import WorkflowClient
        result=WorkflowClient(args.url).setup_proposal(query(decode(raw)))
        print(json.dumps(result,ensure_ascii=False,allow_nan=False))
        return 0
    except (ValueError,OSError,HTTPException) as exc:
        print(json.dumps({'error':str(exc),'generation_submitted':False}),file=sys.stderr)
        return 2


if __name__=='__main__':raise SystemExit(main())
